enclave-runner/src/usercalls/mod.rs

```diff
@@ -330,78 +370,246 @@ impl EnclaveState {
         let last_fd = AtomicUsize::new(fds.keys().cloned().max().unwrap() as _);
 
-        let usercall_ext = usercall_ext.unwrap_or_else( || Box::new(UsercallExtensionDefault));
+        let usercall_ext = usercall_ext.unwrap_or_else(|| Box::new(UsercallExtensionDefault));
+
+        let threads_queue = crossbeam::queue::SegQueue::new();
+
+        for thread in threads_vector {
+            threads_queue.push(Self::event_queue_add_tcs(&mut event_queues, thread));
+        }
 
         Arc::new(EnclaveState {
             kind,
             event_queues,
             fds: Mutex::new(fds),
             last_fd,
             exiting: AtomicBool::new(false),
-            usercall_ext : usercall_ext,
+            usercall_ext,
+            threads_queue,
         })
     }
 
+    fn syscall_loop(
+        &self,
+        io_queue_receive: mpsc::Receiver<UsercallSendData>,
+        work_sender: crossbeam::crossbeam_channel::Sender<Work>,
+    ) -> StdResult<(u64, u64), EnclaveAbort<EnclavePanic>> {
+        while let Ok((coresult, mut state, buf)) = io_queue_receive.recv() {
+            let my_result = match coresult {
+                CoResult::Yield(usercall) => {
+                    let result;
+                    {
+                        let mut input = IOHandlerInput {
+                            enclave: &self,
+                            tcs: &mut state,
+                            work_sender: &work_sender,
+                        };
+                        let (p1, p2, p3, p4, p5) = usercall.parameters();
+                        result = dispatch(&mut Handler(&mut input), p1, p2, p3, p4, p5);
+                    }
+                    match result {
+                        Ok(ret) => {
+                            work_sender
+                                .send(Work {
+                                    tcs: state,
+                                    entry: CoEntry::Resume(usercall, ret),
+                                })
+                                .expect("Work sender couldn't send data to receiver");
+                            continue;
+                        }
+                        Err(EnclaveAbort::Exit { panic: true }) => {
+                            #[cfg(unix)]
+                            trap_attached_debugger(usercall.tcs_address() as _);
+                            Err(EnclaveAbort::Exit {
+                                panic: EnclavePanic::from(buf.into_inner()),
+                            })
+                        }
+                        Err(EnclaveAbort::Exit { panic: false }) => {
+                            match state.mode {
+                                EnclaveEntry::ExecutableNonMain | EnclaveEntry::Library => {
+                                    // If the enclave is in the exit-state, threads are no
+                                    // longer able to be launched
+                                    if !self.exiting.load(Ordering::SeqCst) {
```

Comment:
This doesn't make sense. This is handling `EnclaveAbort::Exit` so if you get here you must be exiting. Remember what I said before about not extracting the TCS from a `Usercall`? Remove the `fn get_tcs` function.